File: leetcode1380/solution.cpp
class Solution {
public:
    vector<int> luckyNumbers (vector<vector<int>>& matrix) {
        /*
        Time: O(m * n) | Space: O(m + n)

        where m is the number of rows and n is the number of columns in the matrix
        */
        vector<int> row(matrix.size(), INT_MAX);
        vector<int> col(matrix[0].size(), INT_MIN);

        vector<int> res;

        for (int i = 0; i < matrix.size(); ++i) {
            for (int j = 0; j < matrix[0].size(); ++j) {
                row[i] = min(row[i], matrix[i][j]);
                col[j] = max(col[j], matrix[i][j]);
            }
        }

        for (int i = 0; i < matrix.size(); ++i) {
            for (int j = 0; j < matrix[0].size(); ++j) {
                if (row[i] == col[j]) {
                    res.push_back(row[i]);
                }
            }
        }

        return res;
    }
};

File: leetcode1380/solution.rs
impl Solution {
    pub fn lucky_numbers (matrix: Vec<Vec<i32>>) -> Vec<i32> {
        /// Time O(m * n) | Space O(m + n)
        /// 
        /// m -> number of rows
        /// n -> number of columns
        let mut row: Vec<i32> = vec![i32::MAX; matrix.len()];
        let mut col: Vec<i32> = vec![i32::MIN; matrix[0].len()];

        let mut res: Vec<i32> = Vec::new();

        for i in 0..matrix.len() {
            for j in 0..matrix[0].len() {
                row[i] = row[i].min(matrix[i][j]);
                col[j] = col[j].max(matrix[i][j]);
            }
        }

        for i in 0..matrix.len() {
            for j in 0..matrix[0].len() {
                if row[i] == col[j] {
                    res.push(row[i]);
                }
            }
        }

        res
    }
}